src/dune/artifact_substitution.ml

```diff
@@ -30,16 +30,22 @@ open Import
 
 type t =
   | Vcs_describe of Path.Source.t
+  | Custom of Path.Build.t
   | Repeat of int * string
 
 let to_dyn = function
   | Vcs_describe p -> Dyn.Variant ("Vcs_describe", [ Path.Source.to_dyn p ])
+  | Custom p -> Dyn.Variant ("Custom", [ Path.Build.to_dyn p ])
   | Repeat (n, s) -> Dyn.Variant ("Repeat", [ Int n; String s ])
 
 let eval t ~get_vcs =
   match t with
   | Repeat (n, s) ->
     Fiber.return (Array.make n s |> Array.to_list |> String.concat ~sep:"")
+  | Custom p ->
+    let f = Path.Build.relative p Custom_build_info.output_file in
+    let s, _ = Build.(contents (Path.build f) |> exec) in
```

Comment:
what happens if the action wrote more than `max_size`?

Reply:
The content will be truncated when artifact subtitution is performed at install-time:

https://github.com/ocaml/dune/blob/c0f1012794c771a70fe2ef83c6854592def2aad7/src/dune/artifact_substitution.ml#L53